### PR TITLE
Handle restart piped command

### PR DESCRIPTION
**Pull request by khanhtc1202**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by khanhtc1202** [comment]
@nghialv @knanao this PR is ready to review, ptal when you guys have time.

File: pkg/app/piped/cmd/piped/piped.go [hunk 1]
```diff
@@ -126,7 +126,11 @@ func NewCommand() *cobra.Command {
 }
 
 func (p *piped) run(ctx context.Context, input cli.Input) (runErr error) {
-	group, ctx := errgroup.WithContext(ctx)
+	// Make a cancellable context.
+	cCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	group, ctx := errgroup.WithContext(cCtx)
```
Comment (nghialv) [inline pkg/app/piped/cmd/piped/piped.go:133]: Why do we need to create a new context with new name?

```suggestion
	ctx, cancel := context.WithCancel(ctx)
	defer cancel()

	group, ctx := errgroup.WithContext(ctx)
```

File: pkg/app/piped/cmd/piped/piped.go [hunk 2]
```diff
@@ -487,6 +491,60 @@ func (p *piped) run(ctx context.Context, input cli.Input) (runErr error) {
...
+			for _, command := range commands {
+				if command.IsRestartPipedCmd() {
+					shouldStop = true
+					stopCmd = command
+				}
+			}
+
+			if !shouldStop {
+				return false, nil
+			}
```
Comment (nghialv) [inline pkg/app/piped/cmd/piped/piped.go:517]:
```suggestion
			var stopCmd    model.ReportableCommand

			for _, command := range commands {
				if command.IsRestartPipedCmd() {
					stopCmd = command
					break
				}
			}

			if stopCmd == nil {
				return false, nil
			}
```

File: pkg/app/piped/cmd/piped/piped.go [hunk 3]
```diff
@@ -487,6 +491,60 @@ func (p *piped) run(ctx context.Context, input cli.Input) (runErr error) {
...
+
+			if err := stopCmd.Report(ctx, model.CommandStatus_COMMAND_SUCCEEDED, nil, []byte(stopCmd.Id)); err != nil {
+				return false, fmt.Errorf("failed to report command %s: %w", stopCmd.Id, err)
+			}
+
+			return true, nil
+		}
+
+		group.Go(func() error {
+			ticker := time.NewTicker(p.gracePeriod)
```
Comment (nghialv) [inline pkg/app/piped/cmd/piped/piped.go:527]: Use a new duration instead of `p.gracePeriod`

File: pkg/app/piped/cmd/piped/piped.go [hunk 4]
```diff
@@ -487,6 +491,60 @@ func (p *piped) run(ctx context.Context, input cli.Input) (runErr error) {
 		})
 	}
 
+	// Check for stop command.
+	{
+		stopCommandHandler := func(ctx context.Context) (bool, error) {
```
Comment (nghialv) [inline pkg/app/piped/cmd/piped/piped.go:496]: nit: Move to a new function to avoid this long block.

File: pkg/app/piped/cmd/piped/piped.go (hunk 1 above)
Comment (khanhtc1202) [reply pkg/app/piped/cmd/piped/piped.go:133]: nice catch 👍

File: pkg/app/piped/cmd/piped/piped.go [hunk 5]
```diff
@@ -487,6 +491,57 @@ func (p *piped) run(ctx context.Context, input cli.Input) (runErr error) {
...
+				select {
+				case <-ticker.C:
+					shouldStop, err := stopCommandHandler(ctx)
+					// Don't return an error to continue this goroutine execution.
+					if err != nil {
+						input.Logger.Error("failed to check/handle piped stop command", zap.Error(err))
+					}
+					if shouldStop {
+						input.Logger.Info("stop piped due to restart piped requested")
+						cancel()
```
Comment (knanao) [inline pkg/app/piped/cmd/piped/piped.go:513]: This sound unnecessary. 👀

Comment (knanao) [reply pkg/app/piped/cmd/piped/piped.go:513]: oops, I misunderstood and this is required, so let's keep this.

Comment (knanao) [reply pkg/app/piped/cmd/piped/piped.go:513]: Sorry, I did not catch your point around this.
If `shouldStop` is true, this run method returns an error, right?

Comment (khanhtc1202) [reply pkg/app/piped/cmd/piped/piped.go:513]: sorry which run method do you say?

Comment (knanao) [reply pkg/app/piped/cmd/piped/piped.go:513]: OK, I totally understand this after talking directly. Thanks

**Review by knanao — COMMENTED** [review]
LGTM

File: pkg/app/piped/cmd/piped/piped.go [hunk 6]
```diff
@@ -801,3 +834,30 @@ func loginToOCIRegistry(ctx context.Context, execPath, address, username, passwo
...
+		}
+	}
+
+	if stopCmd == nil {
+		return false, nil
+	}
+
+	if err := stopCmd.Report(ctx, model.CommandStatus_COMMAND_SUCCEEDED, nil, []byte(stopCmd.Id)); err != nil {
+		return false, fmt.Errorf("failed to report command %s: %w", stopCmd.Id, err)
+	}
```
Comment (nghialv) [inline pkg/app/piped/cmd/piped/piped.go:860]: @khanhtc1202 User could send multiple commands from web console, so how about marking all of them as success before exiting?

Comment (khanhtc1202) [reply pkg/app/piped/cmd/piped/piped.go]: For sure, updated 👍

**Pull request merged**